README
Quick Notes
===========

1. Compile Problems

At this time, you must run the "pre-build" script before you run make.  It is
safe to run before or after "configure".  This is to create some parser files
needed for the build.  Hopefully this step will be removed in version 0.3.

2. Log file

The log file is written by default to log.out which can become quite large. In case
something went wrong (e.g. stellarlegacy crashed) please keep a copy of the log.out
file but do NOT send it to the mailing list (did I mention that log.out is a big 
file? ;-). 

You can disable the log file completly, by commenting out the following line in
common/defs.h (and you have to recompile of course):

#define LOG_ENABLED

Also you can reduce the number of checks and some extra log info if you comment out

#define SECURE_MODE

(the same file). Note that any change in defs.h probably causes a complete recompile
since almost all files depend on defs.h.




ChangeLog
2002-07-12 Marco Krohn <marco.krohn@gmx.de>
 
  * added version information for libraries
  * moved rulesystem.* to gamelogic/
  * modules/ is now a shared library
  * several upper -> lower fixes

2002-07-07 Marco Krohn <marco.krohn@gmx.de>

  * made TAction::fActor a TDoubleLink
  * implemented fleet movement conflict resolution
  * fixed a bug in the fleet movement action generation
  * added a common base class for TDoubleLink

2002-07-04 Marco Krohn <marco.krohn@gmx.de>

  * added some more methods to TDataPackageList (addCopy) which make it easier
    to add information which is only temporarily available (the info is first
    copied and then stored).

2002-07-03 Marco Krohn <marco.krohn@gmx.de>

  * started work on conflict resolution
  * converted the missing lower->upper for actions
  * removed fleetaction.h/.cpp, these are now (thanks to Alberto) automatically
    generated :-)

2002-07-01 Alberto Barsella <ishark@free.fr>

  * Fixes in act-to-cpp to generate good names
  * Modified gamelogic/Makefile.am to be able to automatically generate
    .cpp+.h from .act (there are problems)
  * action types are uppercase, addStandardActions makes sure the action
    get created

2002-06-30 Marco Krohn <marco.krohn@gmx.de>

  * using the prototype pattern framework (prototype.h) for the actions as well
    the only obvious changes:
    
      action.is( ActionType(MOVE_FLEET) )   -->   PMAction.isType( act, "move_fleet" )
      TAction::create("Add Waypoint"))      -->   PMAction.create("add_waypoint")

  * using a validate cache for actions again (the cache only works inside TAction)   

2002-06-26 Alberto Barsella <ishark@free.fr>

  * improved act-to-cpp to actually generate compilable code (does not mean
    that it works :)
  * modified TAction::load/save not to call the specific sub-class
    methods (the subclass will call the parent's)

2002-06-24 Marco Krohn <marco.krohn@gmx.de>

  * eliminated unneeded variable from TAction (ObjectsAreValid?)
  * improved TLog (now using singelton mechanism to guarantee that
    log output is possible in constructors)
  * moved code from fleetaction.cpp => fleetaction.act
    .cpp/.h are now generated from the .act file
    (warning: does not compile at the moment: if you need it, comment the offending lines
    the real fix needs some changes in the act-to-cpp script)

2002-06-24 Alberto Barsella <Alberto.Barsella@univ-lille1.fr>

  * rewritten runSequence loop to delete actions immediately and skip
    over actions which are marked as executed

2002-06-23 Alberto Barsella <Alberto.Barsella@univ-lille1.fr>

  * added act-to-cpp tool and (outdated, for the moment) fleetaction.act
  * some fixes around to add needed using namespace std; for gcc3 
    compatibility
  * fix in TActionSequencer::runSequence to actually call resolveConflict

2002-06-23 Marco Krohn <marco.krohn@gmx.de>

  * made fObjects, fObservers in action a TPEList which fixes a memory leak
    and makes loading/saving easier.
  * added method in action to add "involved" action objects
  * bug fix in TActionSequencer: avoided infinite loop in case of conflicts

2002-06-22 Marco Krohn <marco.krohn@gmx.de>

  * renamed a few methods in TDataPackageList, TRuleSystem
  * improved the macros for creating rule atoms
  * rewrote documentation of rule atoms
  * extended the prototype pattern a bit. Each object now also stores
    a type ID and a type name. This is needed for the action which at
    the moment uses its own prototype system => will change this now...

2002-06-07 Marco Krohn <marco.krohn@gmx.de>

  * activated the move fleet rule atom => fleets do move now
  * extended TRuleAtom: it is now possible to send different data to
    the rule atom
  * improved the fleet movement rule:
    - fleets can reach a waypoint

2002-06-04 Marco Krohn <marco.krohn@gmx.de>

  * some major cleanups and changes in rulesystem
    now making use of TPrototypeManager

2002-06-03 Marco Krohn <marco.krohn@gmx.de>

  * move actions are added to all stobj
  * added prototype.h to /common
    => this makes implementing the prototype pattern easy
  * ruleatoms are now implemented in a way that they make use of the prototype pattern

2002-05-14 Alberto Barsella <Alberto.Barsella@univ-lille1.fr>

  * modified TActionSequencer to use multiple lists and associated stuff
    in TAction
  * fixed some upper/lowecase: getType/TypeName in TType

2002-05-12 Marco Krohn <marco.krohn@gmx.de>

  * moved action sequencer from TGame -> TGameSC
  * moved method "initializePlayerOrders" from TGameSC -> TGameClient
  * client can now reload the old actions :-)
    (just try giving some actions to your fleet, quit, and start in the client modus again)
  * the server mode does work => actions are executed :-)

2002-05-11 Marco Krohn <marco.krohn@gmx.de>

  * two smaller bug fixes, one in the stringtools, one in TGame
  * fixed another bug, caused by a missing copy constructor in TWaypoint
  * reverted changes in TFleet: getType->GetType since there "parents" are already
    written in uppercase and only changing the childs breaks of course everything.
  * added the most important race (the owner of almost everything: Iluvatar)
    => all(??) objects have now an owner and a creator
  * some smaller extensions in TPlayer and TRace
  * implemented "reset" method in TLinkUpdate

2002-05-06 Marco Krohn <marco.krohn@gmx.de>

  * TRuleSystem is now a singelton
  * some minor changes in TRuleSystem
  * removed the galaxy and rulesystem entries from TGame
    (both can be accessed by instance() via TRuleSystem / TUniverse)
  * some modifications in TPList (it now compiles even with -pedantic)
  * added & modified several methods in TPList
    => there is now a goTo() method
  * implemented operator== for TWaypoint
  * TUniverse uses a "list" instead of a "map" now
  * seperated some game code => there are four TGame classes now:
    TGameServer, TGameCreate, TGameSC, TGameClient
  * fixed several warnings in action.cpp

stellarlegacy/sl.cpp
// $Id: sl.cpp,v 1.45 2002/07/12 10:08:02 mkrohn5 Exp $

// Main File
// Written by: Marco Krohn

// Copyright (C) 1999 - 2001, Marco Krohn, <marco.krohn@gmx.de>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 1, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

#include <iostream>
#include <string>

#include "common/defs.h"
#include "admin/game.h"
#include "admin/game-client.h"
#include "admin/game-create.h"
#include "admin/game-server.h"
#include "admin/player.h"
#include "common/random.h"
#include "gamelogic/galaxy.h"
#include "gamelogic/stobj.h"
#include "gamelogic/fleet.h"
#include "gamelogic/stolist.h"
#include "gamelogic/universe.h"

#include "gamelogic/fleetaction.h"

using namespace std;





//----------------------------------------  Prototypes  ----------------------------------------

int parseArguments( int argc, char* argv[] );
void printUsageDescription();
bool runTestScripts();
void startLocalServer();





//----------------------------------------  Main Function  ----------------------------------------

int main( int argc, char* argv[] )
{
int iReturn;

  cout << "This is Stellar Legacy - Version " << VERSION << endl << endl;

  if ( argc == 1 )  {
    printUsageDescription();
    exit( 0 );
  }

  #ifdef LOG_ENABLED
    LOG("-", "main(int, char*[])") << "This is Stellar Legacy - Version " << VERSION << endl
      << endl
      << "The log system is enabled. The log file is called log.out" << endl;
    cout << "The log system is enabled. The log file is called log.out" << endl;
  #else
    cout << "The log system is disabled." << endl
      << "Note: this simply means that SL will not produce a log file." << endl;
  #endif

  TRandom::initialize();
  LOG("-", "main(int, char*[])") << "Random number init: " << TRandom::getInitNumber() << endl;

  iReturn = parseArguments( argc, argv );

  LOG("-", "main(int, char*[])") << "Main function return value: " << iReturn << endl
    << "---------------------------------------" << endl
    << "Destructor calls from global objects:" << endl;

  return iReturn;
}



int parseArguments( int argc, char* argv[] )
{
int i, iReturn = 0;
string sParameter, sPort, sHost;

  i = 1;
  sParameter = argv[i];
  if ( sParameter=="--testing" || sParameter=="-t" )  {
    if ( runTestScripts() == false )  return 1;
    i++;
    if ( i < argc )  sParameter = argv[i];
  }

  if ( i < argc )  {
    if ( sParameter=="--server" || sParameter=="-s" )  {
      i++;
      if ( i < argc )  sPort = argv[i];
      // Interface to Piers server program
      // ...
    }
    else if ( sParameter=="--client" || sParameter=="-c" )  {
      i++;
      if ( i < argc )  sHost = argv[i];
      i++;
      if ( i < argc )  sPort = argv[i];
      cerr << "Sorry, there is no client at the moment" << endl;
    }
    else if ( sParameter=="--localserver" || sParameter=="-l" )  {
      startLocalServer();
    }
    else  {
      cerr << endl << "unknown parameter: \"" << sParameter << "\""<< endl << endl;
      printUsageDescription();
    }
  }

  return iReturn;
}



void printUsageDescription()
{
  cerr << "usage:" << endl;
  cerr << endl;
  cerr << "sl --testing" << endl;
  cerr << "sl [--testing] --server [port]" << endl;
  cerr << "sl [--testing] --client [host] [port]" << endl;
  cerr << "sl [--testing] --localserver" << endl;
  cerr << endl;
  cerr << "Abbreviations: -c = --client" << endl;
  cerr << "               -l = --localserver" << endl;
  cerr << "               -s = --server" << endl;
  cerr << "               -t = --testing" << endl;
}



bool generateNewGame( const string& gameDir )
{
  bool ret = true;

  cout << "Generating new game...." << endl;
  TGameCreate* create = new TGameCreate( gameDir, "no-one" );

  if ( create->createGame() == true )  {
    cout << "game successfully created!" << endl;
    cout << "saving server master file ..." << endl;
    ret &= create->saveServerMasterFile();
    delete create;                                         // cleaning up

    cout << "and now starting the game server ... " << endl;
    TGameServer* server = new TGameServer( gameDir, "no-one" );
    cout << "load the server master file ... " << endl;
    ret &= server->loadServerMasterFile();
    cout << "creating player data files ..." << endl;
    ret &= server->savePlayerDataFiles();
    delete server;
  }
  else  {
    delete create;
    ret = false;
  }

  return ret;
}



bool playTurnFleetMenu( TGameClient* client )
{
TStellarObjectList SOList;
TStellarObjectList::iterator it;
int4 num, selected;
char cInput;
TAction* action;

  cout << "AVAILABLE FLEETS:" << endl;

  // get list of fleets owned by the player
  // we simply use the first galaxy here (note that an universe can consist of more than one galaxy)
  TGalaxy* galaxy = *(TUniverse::instance()->getGalaxies()).begin();

  if ( galaxy->getStellarObjectList().Filter(&SOList, TSOLFilterAND(TSOLFilterType(TType::FLEET), TSOLFilterController(*client->getActivePlayer()))) )
  {
    // show list of fleets to player and ask for the fleet #
    for (num = 0, it = SOList.begin(); it != SOList.end(); ++it, num++)
      cout << num << ". " << (*it)->getName() << "  id=" << (*it)->getID() << endl;

    if (num == 1)
      selected = 1;
    else
      do {
        cout << "Fleet #";
        cin >> selected;
      } while (selected < 0 || selected > num);

    // fetch pointer
    for (it = SOList.begin(); selected; ++it, selected--);
    TFleet& myfleet = *((TFleet*) *it);

    do {
      // show info about fleet
      cout << endl << "SELECTED FLEET: " << myfleet.getName()
           << " at position " << myfleet.getPos() << endl;

      // ask for action to be performed
      cout << endl << "<l>ist waypoints, <s>et waypoint (q)uit" << endl << "> ";
      cin >> cInput;

      // only action: set waypoint, ask for destination
      action = 0;
      switch(cInput) {
        case 'l':
          // fetch the waypoints and print them
          if (myfleet.hasWaypoint()) {
            const TWaypointList* wl = myfleet.getWaypointList();
            TWaypointList::const_iterator wi;

            for (wi = wl->begin(); wi != wl->end(); ++wi) {
              if ((*wi)->isStellarObject())
                cout << "Go to: " << (*wi)->getStellarObject()->getName() << endl;
              else
                cout << "Go to position " << (*wi)->getPos() << endl;
            }
          } else
            cout << myfleet.getName() << ": empty waypoint list" << endl;
          break;

        case 's':
          if ( !(action = PMAction.create("ADD_WAYPOINT")) )
            Fatal("-Client-", "PlayTurnFleetMenu", "unable to create a 'add waypoint' action", -1);

          double dx, dy;
          int2 warp;
          cout << "Enter destination as x y (NO COMMA)" << endl << "> ";
          cin >> dx >> dy;
          // check range blah blah

          cout << "Enter warp speed" << endl << "> ";
          cin >> warp;                                    
          // check range blah blah

          // generate the action
          TActionAddWaypoint* actionWP = dynamic_cast<TActionAddWaypoint*>(action);
          actionWP->setCreator( myfleet.getOwner() );
          actionWP->setActor( &myfleet );
          actionWP->setWaypoint( TWaypoint(TPoint(dx,dy),warp) );

          break;
      }

      if (action) {
        // add action to Game's action sequencer
        // (which implies execute action)
        if ( client->recordPlayerAction(action) == false )
          cout << "ERROR executing the action!!!! (try again?)" << endl;
        else
          cout << "Ok, action was executed just fine." << endl;
      }
    } while (cInput != 'q');

  }
  else  {
    cout << client->getActivePlayer()->getName() << ", you don't have any fleets!" << endl;
  }

  return true;
}



bool playTurn(const string& gameDir, const string& playerName )
{  
bool ret = false;
bool finished = false;
char cInput;


  cout << "Loading turn file...." << endl;
  TGameClient* client = new TGameClient( gameDir, playerName );

  if ( client->loadServerMasterFile() == true)  {
    if ( client->initializePlayerOrders(playerName) == true )  {
      ret = true;
      cout << "Playing turn...." << endl;

      while (!finished) {
        cout << "(f)leet orders ..... (s)ave and quit or just (q)uit" << endl
             << "> ";
        cin >> cInput;

        switch(cInput) {
          case 'f':
            playTurnFleetMenu( client );
            break;
          case 's':
            cout << "Saving orders....." << endl;
            client->savePlayerOrderFile();
            finished = true;
            break;
          case 'q':
            finished = true;
            break;
        }
      }
    }
    else {                                                 // could not initializePlayerOrders
      cout << "could not init player, please check if " << playerName
        << " is really a player in this game" << endl;
    }
  }

  delete client;
  return ret;
}



bool generateTurn( const string& gameDir )
{
bool ret = false;

  TGameServer* server = new TGameServer( gameDir, "server" );

  if ( server->loadServerMasterFile() == true )  {
    cout << "loading player orders file ... " << endl;
    server->loadPlayerOrderFiles();                        // loads all order files
    cout << "backup of the game files ... " << endl;
    server->backupGameFiles();                             // create a backup copy of the files

    server->incrementTurn();
    cout << "executing the actions ... " << endl;
    server->executeActions();

    cout << "saving the server master file ... " << endl;
    server->saveServerMasterFile();
    cout << "saving the player data files ... " << endl;
    server->savePlayerDataFiles();

    ret = true;
  }

  delete server;

  return ret;
}



void startLocalServer()
{
  char cInput;
  string sGameDir;
  string sPlayerName;
  bool finished = false;

  cout << endl << "Enter your player name [default = Marco]" << endl << "> ";
  getline(cin, sPlayerName, '\n');
  if (!sPlayerName.length())  sPlayerName = "Marco";
//    sPlayerName = "Test Player";


  cout << endl << "Enter game identificator/directory [default = testgame]" << endl << "> ";
  getline(cin, sGameDir, '\n');
  if (!sGameDir.length())
    sGameDir = "testgame";

  while (!finished) {
    cout << endl << "GAME DIRECTORY: " << sGameDir << endl 
         << "PLAYER: " << sPlayerName << endl;
    cout << "(l)oad server master file and generate turn   (server)" << endl
         << "(p)lay turn                                   (client)" << endl
         << "(c)reate game                                 (create)" << endl
         << "(q)uit" << endl << "> ";
    cin >> cInput;

    switch(cInput) {
      case 'l':
        if ( generateTurn(sGameDir) == true )
          cout << "Turn successfully generated!" << endl;
        else
          cout << "ERROR generating turn!" << endl;
        break;

      case 'p':
        if ( playTurn(sGameDir, sPlayerName) == true )
          cout << "Turn successfully played!" << endl;
        else
          cout << "ERROR playing turn! (make sure your player name is in the game)" << endl;
        break;

      case 'c':
        if ( generateNewGame(sGameDir) )
          cout << "Game successfully generated!" << endl;
        else
          cout << "ERROR generating game! (make sure that GAME DIR/universe-definition.sl exists)" << endl;
          break;

      case 'q':
        finished = true;
    }
  }
}





/*
* $Log: sl.cpp,v $
* Revision 1.45  2002/07/12 10:08:02  mkrohn5
* *** empty log message ***
*
* Revision 1.44  2002/07/08 06:28:58  mkrohn5
* TAction_XXX -> TActionXXX
*
* Revision 1.43  2002/07/03 11:59:39  mkrohn5
* - ActionAddWa.. -> Action_AddWa...
* - change in makefile.am: testing is compiled last
*
* Revision 1.42  2002/07/02 23:19:25  mkrohn5
* compile fix
*
* Revision 1.41  2002/07/01 20:52:23  ishark
* fixes in act-to-cpp and related stuff
*
* Revision 1.40  2002/07/01 00:09:53  mkrohn5
* "grammer" changes due to changed structure in action.cpp/.h
*
* Revision 1.39  2002/06/28 10:12:02  mkrohn5
* - Add -> add (TDataPackageList)
* - readded validated cache flag in TAction
*
* Revision 1.38  2002/06/22 23:15:37  mkrohn5
* bug fix - fleet selection did not work correctly
*
* Revision 1.37  2002/06/21 08:31:52  mkrohn5
* changed method name in TDatapackageList:
* Get -> getAndDelete
*
* Revision 1.36  2002/05/13 23:03:30  mkrohn5
* removed debug output
* activated server
*
* Revision 1.35  2002/05/12 01:17:24  mkrohn5
* - moved some methods around
* - the client now is able to load the old actions
*
* Revision 1.34  2002/05/11 23:59:58  mkrohn5
* minor modifications
*
* Revision 1.33  2002/05/11 10:40:42  mkrohn5
* more bugfixes
*
* Revision 1.32  2002/05/10 23:46:08  mkrohn5
* edited some strings
*
* Revision 1.31  2002/05/10 10:41:22  mkrohn5
* number of changes due to changed TGame class
*
* Revision 1.30  2002/04/01 21:43:30  mkrohn5
* bugfix which prevented a call to saveAttributes
*
* Revision 1.29  2002/04/01 19:20:10  mkrohn5
* made it a lot easier to add data to an action
*
* Revision 1.28  2002/03/24 18:06:51  mkrohn5
* uppercase -> lowercase fixes
*
* Revision 1.27  2001/12/16 11:37:22  ishark
* TLink now uses dynamic_cast<>, lots of TAction-related fixes
*
* Revision 1.26  2001/12/09 13:25:41  ishark
* minor fixes in game.cpp and sl.cpp
*
* Revision 1.25  2001/12/07 00:21:15  mkrohn5
* - converted galaxy to the new style
* - galaxy now uses the stellar object list (the code is now a lot more readable :-)
* - made TStellarObjectList a list instead of a deque (makes list operations a lot faster)
*
* Revision 1.24  2001/12/05 00:40:47  mkrohn5
* - some conversions XXX -> xXX
* - implemented the basic rule "fleet movement"
*
* Revision 1.23  2001/12/04 00:37:50  mkrohn5
* ooops, that actually should go in yesterday :-(
*
* Revision 1.22  2001/12/02 14:33:45  mkrohn5
* NULL -> 0
*
* Revision 1.21  2001/11/08 21:13:54  ishark
* action uses new method names, additions in sl/race/game
*
* Revision 1.20  2001/11/06 18:14:18  mkrohn5
* removed unnecessary include
*
* Revision 1.19  2001/11/04 02:18:31  mkrohn5
* moved some files around
*
* Revision 1.18  2001/11/02 02:12:07  mkrohn5
* fixed possible crash (see global.cpp, logout for details)
*
* Revision 1.17  2001/10/31 20:23:52  mkrohn5
* some smaller corrections due to Log
*
* Revision 1.16  2001/08/03 19:42:06  ishark
* fixed function/file names to reflect RFCLoadSave
*
* Revision 1.15  2001/07/31 00:32:50  mkrohn5
* replaced tabs by spaces
*
* Revision 1.14  2001/07/28 15:23:11  ishark
* added filters in stolist, more client code in sl.cpp
*
* Revision 1.13  2001/07/25 19:12:31  ishark
* major additions in game/sl.cpp - skeleton of localserver
*
* Revision 1.12  2001/07/10 17:57:32  ishark
* rewrite in TGame (ProcessDataBlock/game creation, changes in TAction
*
* Revision 1.11  2001/07/03 01:03:49  mkrohn5
* added log section at eof
*
*/
